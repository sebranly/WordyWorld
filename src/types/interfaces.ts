export interface Word {
  en: string;
  fr: string;
}
