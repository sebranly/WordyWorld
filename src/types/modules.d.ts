declare module "emoji-from-text" {
  function emojiFromText(text: string, onlyOneMatch?: boolean): any;
  export = emojiFromText;
}
