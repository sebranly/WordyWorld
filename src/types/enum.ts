enum Screen {
  About,
  Explore,
  Game
}

enum WordConnection {
  Addition,
  Anagram,
  Deletion,
  Replacement,
  Same
}

export { Screen, WordConnection };
