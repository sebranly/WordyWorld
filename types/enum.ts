enum Screen {
  About,
  Config,
  Explore
}

export { Screen };
