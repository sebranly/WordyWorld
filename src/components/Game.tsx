// Vendor
import * as React from "react";
import { Col, Grid, Row } from "react-native-easy-grid";
import { Container } from "native-base";
import { StyleSheet, View } from "react-native";
import { Text } from "react-native";

// Internal
import { findWordConnections } from "../helpers/strings";
import { getWordScore } from "../helpers/score";
import { Word } from "../types/interfaces";
import { WordConnection } from "../types/enum";
import { GAME_ROWS, SPECTATOR_INTERVAL } from "../config/settings";

export interface GameProps {
  allWords: Word[];
  initialWordIndex: number;
  style?: any;
}

const Game: React.FC<GameProps> = (props) => {
  // TODO: move out
  // words is the history of all played words. Last one is current one.
  const findReplacements = (words: Word[]) => {
    const word = words[words.length - 1];

    const bannedWords = words.map((w) => w.englishWord);
    const filteredWords = allWords.filter((w) => {
      return !bannedWords.includes(w.englishWord);
    });

    const replacements = filteredWords.filter((w) => {
      const connections = findWordConnections(word.englishWord, w.englishWord);

      const replacementsConnections = connections.filter(
        (c) => c.type === WordConnection.Replacement
      );

      return replacementsConnections.length > 0;
    });

    return replacements;
  };

  // Setup
  const { allWords, initialWordIndex } = props;
  const initialWords = [allWords[initialWordIndex]];

  // Hooks
  const [end, setEnd] = React.useState(false);
  const [score, setScore] = React.useState(0);
  const [words, setWords] = React.useState(initialWords);
  const [wordsConnections, setWordsConnections] = React.useState([] as Word[]);

  // Handlers
  const onSpectator = (initial = false) => {
    /**
     * @external https://stackoverflow.com/questions/53395147/use-react-hook-to-implement-a-self-increment-counter
     * Previous value of the hook state is needed as only first render knows it otherwise
     *
     * TODO: can the setState hooks be cleaned up?
     */
    setWords((words) => {
      const connections = findReplacements(words);

      if (initial) {
        setWordsConnections(connections);
        if (connections.length === 0) setEnd(true);
      } else {
        if (connections.length > 0) {
          const newWord = connections[0];
          const newWords = [...words, newWord];

          const newConnections = findReplacements(newWords);
          setWordsConnections(newConnections);

          const additionalScore = getWordScore(newWord);
          setScore((previousScore) => previousScore + additionalScore);

          if (newConnections.length === 0) setEnd(true);

          return newWords;
        } else {
          setEnd(true);
        }
      }

      return words;
    });
  };

  // Life-cycle
  React.useEffect(() => {
    onSpectator(true);
    const intervalId = setInterval(onSpectator, SPECTATOR_INTERVAL);

    return () => clearInterval(intervalId);
  }, []); // Empty array leads to same behavior as `componentDidMount` (if it was a class)

  const renderRow = (indexRow: number, last?: boolean) => {
    const renderColumn = (letter: string, indexCol: number) => {
      const innerComponent = last ? (
        <View style={styles.viewText}>
          <Text style={styles.text}>{letter}</Text>
        </View>
      ) : null;

      return (
        <Col key={`col-${indexCol}`} style={styles.cell}>
          {innerComponent}
        </Col>
      );
    };

    const renderColumns = () => {
      const word = words[words.length - 1];
      const letters = word.englishWord.split("");

      const columns = letters.map((v, i) => renderColumn(v, i));
      return columns;
    };

    return <Row key={`row-${indexRow}`}>{renderColumns()}</Row>;
  };

  const renderRows = () => {
    const lastIndexRow = GAME_ROWS - 1;
    const allButLastRow = Array.from({ length: lastIndexRow }, (_v, i) =>
      renderRow(i)
    );

    return [...allButLastRow, renderRow(lastIndexRow, true)];
  };

  return (
    <Container>
      <Text style={styles.title}>Game</Text>
      <Grid style={styles.grid}>{renderRows()}</Grid>
      <View style={styles.example}>
        <Text>{`Words: ${allWords.length}`}</Text>
        <Text>{`History: ${words.map((w) => w.englishWord)}`}</Text>
        {!end && (
          <Text>{`Replacement words: ${wordsConnections
            .map((w) => w.englishWord)
            .join(" ")}`}</Text>
        )}
        {<Text>{`Score: ${score}`}</Text>}
        {end && <Text>Game Over</Text>}
      </View>
    </Container>
  );
};

const styles = StyleSheet.create({
  cell: {
    backgroundColor: "grey",
    margin: 2,
  },
  example: {
    backgroundColor: "yellow",
    flex: 1,
  },
  grid: {
    flex: 3,
  },
  text: { fontSize: 30 },
  title: {
    textAlign: "center",
  },
  viewText: {
    alignItems: "center",
    flex: 1,
    justifyContent: "center",
  },
});

export { Game };
