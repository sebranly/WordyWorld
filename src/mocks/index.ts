// Internal
import { Word } from "../types/interfaces";

const mockWords: Word[] = [
  { en: "car", fr: "voiture" },
  { en: "cat", fr: "chat" }
];

export { mockWords };
