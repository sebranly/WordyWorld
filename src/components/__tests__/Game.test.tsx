// Vendor
import React from "react";
import { render } from "@testing-library/react-native";

// Internal
import { Game } from "../Game";
import { mockWords } from "../../mocks";
import { Word } from "../../types/interfaces";

it("renders", () => {
  const props = { words: mockWords };

  const { container } = render(<Game {...props} />);
  expect(container.children.length).toBeGreaterThan(0);
  expect(container.children).toMatchSnapshot();
});
