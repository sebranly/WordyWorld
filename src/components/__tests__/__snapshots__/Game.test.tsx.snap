// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
Array [
  <View
    style={
      Object {
        "backgroundColor": "#fff",
        "flex": 1,
        "height": 1334,
      }
    }
  >
    <Text
      style={
        Object {
          "textAlign": "center",
        }
      }
    >
      Game
    </Text>
    <View
      style={
        Object {
          "flex": 3,
          "flexDirection": "column",
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "grey",
              "flex": 1,
              "flexDirection": "column",
              "margin": 2,
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "grey",
              "flex": 1,
              "flexDirection": "column",
              "margin": 2,
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "grey",
              "flex": 1,
              "flexDirection": "column",
              "margin": 2,
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "grey",
              "flex": 1,
              "flexDirection": "column",
              "margin": 2,
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "grey",
              "flex": 1,
              "flexDirection": "column",
              "margin": 2,
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "grey",
              "flex": 1,
              "flexDirection": "column",
              "margin": 2,
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "grey",
              "flex": 1,
              "flexDirection": "column",
              "margin": 2,
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "grey",
              "flex": 1,
              "flexDirection": "column",
              "margin": 2,
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "grey",
              "flex": 1,
              "flexDirection": "column",
              "margin": 2,
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "grey",
              "flex": 1,
              "flexDirection": "column",
              "margin": 2,
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "grey",
              "flex": 1,
              "flexDirection": "column",
              "margin": 2,
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "grey",
              "flex": 1,
              "flexDirection": "column",
              "margin": 2,
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "grey",
              "flex": 1,
              "flexDirection": "column",
              "margin": 2,
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "grey",
              "flex": 1,
              "flexDirection": "column",
              "margin": 2,
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "grey",
              "flex": 1,
              "flexDirection": "column",
              "margin": 2,
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "grey",
              "flex": 1,
              "flexDirection": "column",
              "margin": 2,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "justifyContent": "center",
              }
            }
          >
            <Text
              style={
                Object {
                  "fontSize": 30,
                }
              }
            >
              c
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "backgroundColor": "grey",
              "flex": 1,
              "flexDirection": "column",
              "margin": 2,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "justifyContent": "center",
              }
            }
          >
            <Text
              style={
                Object {
                  "fontSize": 30,
                }
              }
            >
              a
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "backgroundColor": "grey",
              "flex": 1,
              "flexDirection": "column",
              "margin": 2,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "justifyContent": "center",
              }
            }
          >
            <Text
              style={
                Object {
                  "fontSize": 30,
                }
              }
            >
              r
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "backgroundColor": "yellow",
          "flex": 1,
        }
      }
    >
      <Text>
        Words: 2
      </Text>
      <Text>
        History: car
      </Text>
      <Text>
        Replacement words: cat
      </Text>
      <Text>
        Score: 0
      </Text>
    </View>
  </View>,
]
`;
